rust/plugin-lib/src/state_cache.rs

```diff
@@ -0,0 +1,517 @@
+                        } else {
+                            ctx.truncate_cache(start);
+                        }
+                    } else {
+                        ctx.state.chunk.clear();
+                        ctx.state.chunk_offset = 0;
+                        ctx.truncate_cache(start);
+                    }
+                } else {
+                    ctx.state.chunk.clear();
+                    ctx.state.chunk_offset = 0;
+                    ctx.truncate_cache(start);
+                }
+                self.handler.update(ctx);
+                Some(Value::from(0i32))
+            }
+            PluginRequest::DidSave { ref path } => {
+                ctx.state.path = Some(path.to_owned());
+                self.handler.did_save(ctx);
+                None
+            }
+        }
+    }
+
+    fn idle(&mut self, peer: plugin_base::PluginCtx, token: usize) {
+        let ctx = PluginCtx {
+            state: &mut self.state,
+            peer: peer,
+        };
+        self.handler.idle(ctx, token);
+    }
+}
+
+pub fn mainloop<H: Handler>(handler: &mut H) {
+    let mut my_handler = MyHandler {
+        handler: handler,
+        state: MyState::default(),
+    };
+    plugin_base::mainloop(&mut my_handler);
+}
+
+impl<'a, S: Default + Clone> PluginCtx<'a, S> {
+
+    pub fn get_path(&self) -> Option<&PathBuf> {
+        match self.state.path {
+            Some(ref p) => Some(p),
+            None => None,
+        }
+    }
+
+    pub fn add_scopes(&self, scopes: &Vec<Vec<String>>) {
+        self.peer.add_scopes(&self.state.view_id, scopes)
+    }
+
+    pub fn update_spans(&self, start: usize, len: usize, spans: &[ScopeSpan]) {
+        self.peer.update_spans(&self.state.view_id, start, len, self.state.rev, spans)
+    }
+
+    /// Determines whether an incoming request (or notification) is pending. This
+    /// is intended to reduce latency for bulk operations done in the background.
+    pub fn request_is_pending(&self) -> bool {
+        self.peer.request_is_pending()
+    }
+
+    /// Schedule the idle handler to be run when there are no requests pending.
+    pub fn schedule_idle(&mut self, token: usize) {
+        self.peer.schedule_idle(token);
+    }
+
+    /// Find an entry in the cache by line num. On return `Ok(i)` means entry
+    /// at index `i` is an exact match, while `Err(i)` means the entry would be
+    /// inserted at `i`.
+    fn find_line(&self, line_num: usize) -> Result<usize, usize> {
+        self.state.state_cache.binary_search_by(|probe| probe.line_num.cmp(&line_num))
+    }
+
+    /// Find an entry in the cache by offset. Similar to `find_line`.
+    fn find_offset(&self, offset: usize) -> Result<usize, usize> {
+        self.state.state_cache.binary_search_by(|probe| probe.offset.cmp(&offset))
+    }
+
+    /// Get state at or before given line number. Returns line number, offset,
+    /// and user state.
+    pub fn get_prev(&self, line_num: usize) -> (usize, usize, S) {
```

Comment:
The documentation or the name of this function could more clearly indicate that it returns the existing preceding cache entry nearest to `line_num`.

that it returns the last entry which exists in the cache and is <= line_num.

This is a bit nitty, the code is clear.

Reply:
Done.